crates/bevy_ecs/src/query/builder.rs

```diff
@@ -0,0 +1,422 @@
+    /// on an empty builder and then adds all accesses from that builder to self.
+    ///
+    /// Primarily used when inside a [`Self::or`] closure to group several terms.
+    pub fn and(&mut self, f: impl Fn(&mut QueryBuilder)) -> &mut Self {
+        let mut builder = QueryBuilder::new(self.world);
+        f(&mut builder);
+        let access = builder.access().clone();
+        self.extend_access(access);
+        self
+    }
+
+    /// Takes a function over mutable access to a [`QueryBuilder`], calls that function
+    /// on an empty builder, all accesses added to that builder will become terms in an or expression.
+    ///
+    /// ```
+    /// # use bevy_ecs::prelude::*;
+    /// #
+    /// # #[derive(Component)]
+    /// # struct A;
+    /// #
+    /// # #[derive(Component)]
+    /// # struct B;
+    /// #
+    /// # let mut world = World::new();
+    /// #
+    /// QueryBuilder::<Entity>::new(&mut world).or(|builder| {
+    ///     builder.with::<A>();
+    ///     builder.with::<B>();
+    /// });
+    /// // is equivalent to
+    /// QueryBuilder::<Entity>::new(&mut world).push::<Or<(With<A>, With<B>)>>();
+    /// ```
+    pub fn or(&mut self, f: impl Fn(&mut QueryBuilder)) -> &mut Self {
+        let mut builder = QueryBuilder::new(self.world);
+        builder.or = true;
+        builder.first = true;
+        f(&mut builder);
+        self.access.extend(builder.access());
+        self
+    }
+
+    /// Returns a reference to the the [`FilteredAccess`] that will be provided to the built [`Query`].
+    pub fn access(&self) -> &FilteredAccess<ComponentId> {
+        &self.access
+    }
+
+    /// Transmute the existing builder adding erquired accesses.
+    /// This will maintain all exisiting accesses.
+    ///
+    /// If including a filter type see [`Self::transmute_filtered`]
+    pub fn transmute<NewQ: WorldQuery>(&mut self) -> &mut QueryBuilder<'w, NewQ> {
+        self.transmute_filtered::<NewQ, ()>()
+    }
+
+    /// Transmute the existing builder adding erquired accesses.
+    /// This will maintain all existing accesses.
+    pub fn transmute_filtered<NewQ: WorldQuery, NewF: ReadOnlyWorldQuery>(
+        &mut self,
+    ) -> &mut QueryBuilder<'w, NewQ, NewF> {
+        let mut fetch_state = NewQ::init_state(self.world);
+        let filter_state = NewF::init_state(self.world);
+
+        NewQ::set_access(&mut fetch_state, &self.access);
+
+        let mut access = FilteredAccess::default();
+        NewQ::update_component_access(&fetch_state, &mut access);
+        NewF::update_component_access(&filter_state, &mut access);
+
+        self.extend_access(access);
+        // SAFETY:
+        // - We have included all requried acceses for NewQ and NewF
+        // - The layout of all QueryBuilder instances is the same
+        unsafe { std::mem::transmute(self) }
+    }
+
+    /// Create a [`QueryState`] with the accesses of the builder.
+    pub fn build(&mut self) -> QueryState<Q, F> {
```

Comment:
Semantically, it feels like this should take `self` by value instead.

Reply:
The mutating effects are particularly unclear.

Reply:
I agree, but then all the other methods also have to take self by value (which is bad because now you can't pass it around by reference to build with) or you need to split building into multiple lines which is a bit ugly (similar to why App::run takes &mut self).

Happy to change it though.

Reply:
The reason it needs to be mutable is it use the underlying &mut World when initializing the QueryState, although I could use the new `get_state` and only take &self for build since you shouldn't be creating any components when building from a `QueryBuilder`

Reply:
Ah, okay :) I'm happy to leave this as is, but perhaps we could document that for users.